rust/kernel/cpu.rs

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! CPU utilities for SMP systems.
+//!
+//! Note: This module is a work in progress.
+
+use crate::{bindings, error::code, prelude::*};
+use core::{marker::PhantomData, mem::ManuallyDrop};
+
+/// Represents an SMP CPU by its zero-based kernel processor id.
+///
+/// Cpu values can be safely passed between threads (and across physical CPUs) as they are
+/// merely an integer index representing a CPU. All actual interaction with CPUs is late-
+/// bound.
+pub struct Cpu(i32);
+
+struct FnCall<F, R>
```

Comment:
Given that `FnCall` and `call_fn` are only used in `Cpu::call`, please move them inside that function; this avoids namespace pollution and makes the relevant code more local.

Reply:
Given that I'm hoping to add additional SMP functions soon that will use the same primitives, I'm going to leave things as they are unless there's a very strong objection.